src/main/scala/shine/C/AST/Types.scala

```diff
@@ -2,8 +2,9 @@ package shine.C.AST
 
 import arithexpr.arithmetic._
 import shine.C
+import shine.cuda.AST.FragmentType
 
-sealed abstract class Type(val const: Boolean) {
+abstract class Type(val const: Boolean) {
```

Comment:
It is a big downside to not having the Type superclass sealed: pattern matching will not be checked for exhaustiveness.
We should avoid this change.

Reply:
So the problem is which Type should the new `FragmentType` extend?
As mentioned here: https://github.com/rise-lang/shine/pull/100#discussion_r573562914 `BasicType` does not really make sense for `FragmentType` cause a `FragmentType` represents a matrix.
Also there are some problems when extend  `BasicType`, because when the `FragmentType` is created the matrixlayout is still not infered (https://github.com/rise-lang/shine/issues/111).
So currently the `FragmentType` directly extend `Type` which would require to remove the sealed because the FragmentType is a CUDA specific type which is defined in cuad/AST/Types.scala

Reply:
Maybe you could for now put this type in the core C types, it should not be there but it allows to keep the trait sealed and can be fixed later?

Reply:
We will have to revisit the implementation of the type hierarchy at some point. `VectorType` should also not be in the core typing, but it is currently there. The typing hierarchy is currently just not setup for extensions.

I agree with @Bastacyclop's suggestion. We will have to revisit this later.